Fix some assertions when using a canvas without a layer with webgpu

achristensen07 opened the pull request:
#### f816a1743b89245484a0ea3462c76fe88d593d3e
<pre>
Fix some assertions when using a canvas without a layer with webgpu
<a href="https://bugs.webkit.org/show_bug.cgi?id=270705">https://bugs.webkit.org/show_bug.cgi?id=270705</a>
<a href="https://rdar.apple.com/124213314">rdar://124213314</a>

Reviewed by Mike Wyrzykowski.

The test in this PR hit several issues:
1. PresentationContextIOSurface::present asserted, which I changed to an early return.
2. I handle errors when creating objects by returning nullptr instead of an object that
   has no corresponding object in the GPU process, and propogate those errors to JS.
3. I added a WeakPtr check in the lambda in GPUCanvasContextCocoa::prepareForDisplay
4. In order to keep Windows building after this change, I needed to change
   FontPlatformData::Attributes to FontPlatformDataAttributes in WebKit.

* LayoutTests/fast/webgpu/use-canvas-without-layer-expected.txt: Added.
* LayoutTests/fast/webgpu/use-canvas-without-layer.html: Added.
* Source/WebGPU/WebGPU/PresentationContextIOSurface.mm:
* Source/WebKit/Platform/IPC/StreamServerConnection.cpp:
(IPC::StreamServerConnection::dispatchStreamMessages):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteTextureProxy.cpp:
(WebKit::WebGPU::RemoteTextureProxy::~RemoteTextureProxy):
(WebKit::WebGPU::RemoteTextureProxy::createView):
(WebKit::WebGPU::RemoteTextureProxy::destroy):
(WebKit::WebGPU::RemoteTextureProxy::setLabelInternal):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteTextureProxy.h:

Canonical link: <a href="https://commits.webkit.org/276005@main">https://commits.webkit.org/276005@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/24c7b3bb74cec914da456438e84b7e7736602ea5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43343 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22373 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45753 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45978 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39469 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/45647 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26124 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19791 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35831 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43917 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19391 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37335 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16810 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/43214 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/16972 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38394 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1399 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39523 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38713 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47522 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18257 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15023 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42617 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19794 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/37624 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41275 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19973 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5919 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19425 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteTextureProxy.cpp:
```diff
@@ -44,24 +44,24 @@ RemoteTextureProxy::RemoteTextureProxy(RemoteGPUProxy& root, ConvertToBackingCon
-    auto sendResult = send(Messages::RemoteTexture::Destruct());
-    UNUSED_VARIABLE(sendResult);
+    if (m_createdView) {
+        auto sendResult = send(Messages::RemoteTexture::Destruct());
```
why do we only call destruct if `m_createdView` is true?

mwyrzykowski replied:
I see the comment in the commit log but I don't completely follow:
> Messages::RemoteTexture::Destroy was being sent without
   Messages::RemoteTexture::CreateView having been sent

but the texture is created from RemoteDevice::CreateTexture, not RemoteTexture::CreateView

mwyrzykowski commented on Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteTextureProxy.cpp:
```diff
@@ -70,14 +70,18 @@ Ref<WebCore::WebGPU::TextureView> RemoteTextureProxy::createView(const std::opti
 {
-    auto sendResult = send(Messages::RemoteTexture::SetLabel(label));
-    UNUSED_VARIABLE(sendResult);
+    if (m_createdView) {
```
I don't think this is correct because we should be able to set the label on a GPUTexture before creating its view

mwyrzykowski reviewed (changes requested):
I think this patch conflates the concepts of texture views and textures which are not the same thing

mwyrzykowski commented on Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteTextureProxy.cpp:
```diff
@@ -48,30 +48,32 @@ RemoteTextureProxy::~RemoteTextureProxy()
 {
-    auto sendResult = send(Messages::RemoteTexture::Destroy());
-    UNUSED_VARIABLE(sendResult);
+    if (m_createdView) {
```
it should be possible to destroy a texture which doesn't create a view

mwyrzykowski reviewed (changes requested):
I think it is incorrect to prevent GPUTexture.destroy on a texture which has not created any views, as it is not required to create a view to use a GPUTexture

kkinnunen-apple commented on Source/WebKit/Platform/IPC/StreamServerConnection.cpp:
```diff
@@ -174,13 +175,13 @@ StreamServerConnection::DispatchResult StreamServerConnection::dispatchStreamMes
-            Locker locker { m_receiversLock };
-            ASSERT(m_receivers.isEmpty());
-            return DispatchResult::HasNoMessages;
+            if (skipStreamMessage(decoder)) {
```
Fixing the assertion bug in WebGPU doesn't sound like it'd need the skip implementation..

The idea is that if there is a malformed message in the IPC data stream, the whole data stream is invalid. E.g. you cannot calculate where  the next message is in a invalid data stream.

It is a sender error to send a malformed message, and as such the only fix should be to fix the sender to not send the malformed message.

Also adding nontrivial features to IPC could make use of the unit tests. It'd be good to then maybe consider implementing non-trivial features as separate patches to the other fixes, to ease the review and discussion.

mwyrzykowski replied:
Discussed with @achristensen07 and found a four line change to fix the crash instead

mwyrzykowski reviewed (changes requested):
Can we split the PR into two patches please? One which is limited to fixing the crash:
```
diff --git a/Source/WebGPU/WebGPU/PresentationContextIOSurface.h b/Source/WebGPU/WebGPU/PresentationContextIOSurface.h
index 4a2278ff15e8..c3bcad5093cd 100644
--- a/Source/WebGPU/WebGPU/PresentationContextIOSurface.h
+++ b/Source/WebGPU/WebGPU/PresentationContextIOSurface.h
@@ -62,6 +62,7 @@ private:
     };
     Vector<RenderBuffer> m_renderBuffers;
     RefPtr<Device> m_device;
+    RefPtr<Texture> m_invalidTexture;
     size_t m_currentIndex { 0 };
     id<MTLFunction> m_luminanceClampFunction;
     id<MTLComputePipelineState> m_computePipelineState;
diff --git a/Source/WebGPU/WebGPU/PresentationContextIOSurface.mm b/Source/WebGPU/WebGPU/PresentationContextIOSurface.mm
index cfe4ff92e641..cfecaa97086c 100644
--- a/Source/WebGPU/WebGPU/PresentationContextIOSurface.mm
+++ b/Source/WebGPU/WebGPU/PresentationContextIOSurface.mm
@@ -74,6 +74,7 @@ void PresentationContextIOSurface::configure(Device& device, const WGPUSwapChain
 {
     m_renderBuffers.clear();
     m_currentIndex = 0;
+    m_invalidTexture = Texture::createInvalid(device);
 
     if (descriptor.nextInChain)
         return;
@@ -235,7 +236,7 @@ void PresentationContextIOSurface::present()
 Texture* PresentationContextIOSurface::getCurrentTexture()
 {
     if (m_ioSurfaces.count != m_renderBuffers.size() || m_renderBuffers.size() <= m_currentIndex)
-        return nullptr;
+        return m_invalidTexture.get();
 
     auto& texturePtr = m_renderBuffers[m_currentIndex].luminanceClampTexture;
     if (texturePtr.get()) {
diff --git a/Source/WebKit/GPUProcess/graphics/WebGPU/RemotePresentationContext.cpp b/Source/WebKit/GPUProcess/graphics/WebGPU/RemotePresentationContext.cpp
index ac161851907b..11eabde6aeba 100644
--- a/Source/WebKit/GPUProcess/graphics/WebGPU/RemotePresentationContext.cpp
+++ b/Source/WebKit/GPUProcess/graphics/WebGPU/RemotePresentationContext.cpp
@@ -72,6 +72,7 @@ void RemotePresentationContext::unconfigure()
 void RemotePresentationContext::getCurrentTexture(WebGPUIdentifier identifier)
 {
     auto texture = m_backing->getCurrentTexture();
+    ASSERT(texture);
     if (!texture)
         return;
```
and the other with the rest of these changes?

mwyrzykowski commented on Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
```diff
@@ -143,7 +145,9 @@ void RemoteDevice::createTexture(const WebGPU::TextureDescriptor& descriptor, We
 
     auto texture = m_backing->createTexture(*convertedDescriptor);
-    auto remoteTexture = RemoteTexture::create(texture, m_objectHeap, m_streamConnection.copyRef(), identifier);
+    if (!texture)
```
Since we don't signal back to the web process, I think it is incorrect to not add the remote texture to the objectHeap, since the web process will be left with an incomplete connection.

Instead, can we have `RemoteTexture::create` return a `Ref` instead of a `RefPtr` please?

achristensen07 replied:
Using the interface WebCore::WebGpu::Device for both RemoteDeviceProxy where it can fail and DeviceImpl where it can't won't work then.  This null check currently can't be hit

mwyrzykowski reviewed (changes requested):
I don't think we should allow the `create*` functions to skip calling `m_objectHeap.addObject` as the web process will not be informed that object creation failed. Then we encounter a similar issue as https://bugs.webkit.org/show_bug.cgi?id=270800

mwyrzykowski reviewed (changes requested):
It appears https://gpuweb.github.io/cts/standalone/?q=webgpu:api,validation,encoding,cmds,copyTextureToTexture:copy_with_invalid_or_destroyed_texture:* fails after this change

achristensen07 committed:
Fix some assertions when using a canvas without a layer with webgpu
https://bugs.webkit.org/show_bug.cgi?id=270705
rdar://124213314

Reviewed by Mike Wyrzykowski.

The test in this PR hit several issues:
1. PresentationContextIOSurface::present asserted, which I changed to an early return.
2. I handle errors when creating objects by returning nullptr instead of an object that
   has no corresponding object in the GPU process, and propogate those errors to JS.
3. I added a WeakPtr check in the lambda in GPUCanvasContextCocoa::prepareForDisplay
4. In order to keep Windows building after this change, I needed to change
   FontPlatformData::Attributes to FontPlatformDataAttributes in WebKit.

* LayoutTests/fast/webgpu/use-canvas-without-layer-expected.txt: Added.
* LayoutTests/fast/webgpu/use-canvas-without-layer.html: Added.
* Source/WebGPU/WebGPU/PresentationContextIOSurface.mm:
* Source/WebKit/Platform/IPC/StreamServerConnection.cpp:
(IPC::StreamServerConnection::dispatchStreamMessages):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteTextureProxy.cpp:
(WebKit::WebGPU::RemoteTextureProxy::~RemoteTextureProxy):
(WebKit::WebGPU::RemoteTextureProxy::createView):
(WebKit::WebGPU::RemoteTextureProxy::destroy):
(WebKit::WebGPU::RemoteTextureProxy::setLabelInternal):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteTextureProxy.h:

Canonical link: https://commits.webkit.org/276005@main
